lib/Canonicalize/MakeLocalVariables.cpp

```diff
@@ -78,44 +78,44 @@ bool MakeLocalVariables::runOnFunction(llvm::Function &F) {
   const model::Function *ModelF = llvmToModelFunction(*Model, F);
   revng_assert(ModelF);
   using ModelTypesMap = std::map<const llvm::Value *,
-                                 const model::QualifiedType>;
+                                 const model::UpcastableType>;
   ModelTypesMap KnownTypes = initModelTypes(Cache,
                                             F,
                                             ModelF,
                                             *Model,
-                                            /*PointersOnly=*/false);
+                                            /* PointersOnly = */ false);
 
   for (auto *Alloca : ToReplace) {
     Builder.SetInsertPoint(Alloca);
     llvm::Type *ResultType = Alloca->getType();
 
     // Compute the allocated type for the alloca
-    model::QualifiedType AllocatedType;
-    {
-      std::set<model::QualifiedType> StoredTypes;
-      for (const llvm::Use &U : Alloca->uses()) {
-        if (auto *Store = dyn_cast<llvm::StoreInst>(U.getUser())) {
-          if (Store->getPointerOperandIndex() == U.getOperandNo()) {
-            llvm::Value *Stored = Store->getValueOperand();
-            auto It = KnownTypes.find(Stored);
-            if (It != KnownTypes.end()) {
-              StoredTypes.insert(It->second);
+    model::UpcastableType AllocatedType = nullptr;
```

Comment:
`= nullptr` necessary? Leaks an implementation detail.

Reply:
The intention is to stress that we support the possibility of it remaining as `nullptr` after the following loop.

Reply:
What was the final decision on this matter?

Reply:
I ended up going with your suggestion of `UpcastableType::empty()`.

Reply:
Done.